core/unit_tests.c

```diff
@@ -60,6 +61,210 @@ unit_test_atomic_ops(void);
+        MOVE_TO_ZMM(buffer, 2)
+        MOVE_TO_ZMM(buffer, 3)
+        MOVE_TO_ZMM(buffer, 4)
+        MOVE_TO_ZMM(buffer, 5)
+        MOVE_TO_ZMM(buffer, 6)
+        MOVE_TO_ZMM(buffer, 7)
+#    ifdef X64
+        MOVE_TO_ZMM(buffer, 8)
+        MOVE_TO_ZMM(buffer, 9)
+        MOVE_TO_ZMM(buffer, 10)
+        MOVE_TO_ZMM(buffer, 11)
+        MOVE_TO_ZMM(buffer, 12)
+        MOVE_TO_ZMM(buffer, 13)
+        MOVE_TO_ZMM(buffer, 14)
+        MOVE_TO_ZMM(buffer, 15)
+#        ifdef __AVX512F__
+        MOVE_TO_ZMM(buffer, 16)
+        MOVE_TO_ZMM(buffer, 17)
+        MOVE_TO_ZMM(buffer, 18)
+        MOVE_TO_ZMM(buffer, 19)
+        MOVE_TO_ZMM(buffer, 20)
+        MOVE_TO_ZMM(buffer, 21)
+        MOVE_TO_ZMM(buffer, 22)
+        MOVE_TO_ZMM(buffer, 23)
+        MOVE_TO_ZMM(buffer, 24)
+        MOVE_TO_ZMM(buffer, 25)
+        MOVE_TO_ZMM(buffer, 26)
+        MOVE_TO_ZMM(buffer, 27)
+        MOVE_TO_ZMM(buffer, 28)
+        MOVE_TO_ZMM(buffer, 29)
+        MOVE_TO_ZMM(buffer, 30)
+        MOVE_TO_ZMM(buffer, 31)
+#        endif
+#    endif
+    default: FAIL(); break;
+    }
+}
+
+void
+unit_test_get_zmm_caller_saved()
+{
+    /* XXX i#1312: get_zmm_caller_saved(byte* buf) assumes that there is enough space in
+     * the buffer it's being passed. MCXT_NUM_SIMD_SLOTS does not yet reflect this. Once
+     * this happens, the array size should become MCXT_NUM_SIMD_SLOTS.
+     */
+    if (MCXT_NUM_SIMD_SLOTS == 32) {
+        /* This is a just reminder.*/
+        FAIL();
+    }
+    dr_zmm_t ref_buffer[32];
+    dr_zmm_t get_buffer[32];
+    uint base = 0x78abcdef;
+    for (int regno = 0; regno < proc_num_simd_registers(); ++regno) {
+        for (int dword = 0; dword < MCXT_SIMD_SLOT_SIZE / 4; ++dword) {
+            ref_buffer[regno].u32[dword] = 0;
+            get_buffer[regno].u32[dword] = 0;
+        }
+        base += regno;
+        for (int dword = 0; dword < 16; ++dword) {
+            ref_buffer[regno].u32[dword] = base + dword;
+        }
+        write_zmm_aux(ref_buffer, regno);
+    }
+    get_zmm_caller_saved(get_buffer);
+    for (int regno = 0; regno < proc_num_simd_registers(); ++regno) {
+        print_file(
+            STDERR,
+            "zmm%d: 0x%x 0x%x 0x%x 0x%x 0x%x 0x%x 0x%x 0x%x 0x%x 0x%x 0x%x 0x%x 0x%x "
+            "0x%x 0x%x 0x%x - 0x%x 0x%x 0x%x 0x%x 0x%x 0x%x 0x%x 0x%x 0x%x 0x%x 0x%x "
+            "0x%x 0x%x 0x%x 0x%x 0x%x\n",
+            regno, ref_buffer[regno].u32[0], ref_buffer[regno].u32[1],
+            ref_buffer[regno].u32[2], ref_buffer[regno].u32[3], ref_buffer[regno].u32[4],
+            ref_buffer[regno].u32[5], ref_buffer[regno].u32[6], ref_buffer[regno].u32[7],
+            ref_buffer[regno].u32[8], ref_buffer[regno].u32[9], ref_buffer[regno].u32[10],
+            ref_buffer[regno].u32[11], ref_buffer[regno].u32[12],
+            ref_buffer[regno].u32[13], ref_buffer[regno].u32[14],
+            ref_buffer[regno].u32[15], get_buffer[regno].u32[0], get_buffer[regno].u32[1],
```

Comment:
How about instead calling `dump_buffer_as_bytes(STDERR, &get_buffer[regno], sizeof(get_buffer[regno]), DUMP_RAW|DUMP_DWORD)` for each of these (esp the zmm ones)?  Or looping like dump_mcontext does.  Just seems error-prone and long as-is.

Reply:
Done.